### DISCUSSION TITLE
Code auto-formatting and pre-commit

### DISCUSSION THREAD

**Discussion by akaszynski** [pyvista/pyvista#2178] [Ideas]
I think we should consider incorporating some code auto-formatting like [black](https://github.com/psf/black), especially using [pre-commit](https://pre-commit.com/).

Some reasoning:
- May of the changes made to our repository are white space changes or general formatting changes. Having this done in a pre-commit would alleviate a lot of the nit-picks that improve our code but also obscure more critical comments within PRs.
- We can add in several of our libraries within our `Makefile` into pre-commit and simply use them there.
- PyVista developers can add pre-commit hooks into their git config and avoid pushing code that doesn't pass formatting or spell checking. Pytest probably should be optional since it takes some time to run, but code formatting only takes a moment.

Your thoughts @pyvista/developers?

**Comment by adeak** [comment]
I've been using a local pre-commit hook that runs `flake8` and `codespell` ever since those were recenly made hard checks in CI.

`black` is fairly divisive but most of our code is probably compatible with its opinions. If we start using it the question arises whether
1. we take a deep breath and feed all our code through `black` and add the commit in an [`ignore-revs-file`](https://www.git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt). But [github doesn't support this yet](https://github.com/github/feedback/discussions/5033).
2. we only enable it for code touched by PRs.

Speaking as someone who's the most likely to obsess about trivial style changes, `black` might help :P

As for pre-commit, I didn't raise them as an official suggestion because each additional step we try to force on contributors raises the barrier to entry. So anything we have there should be carefully optional. Server-side hooks would be ideal for automatically fixable problems, but github also doesn't support server-side hooks as far as I know.

**Comment by tkoyama010** [comment]
+1 Forgetting to run a format check is always annoying to me.

**Comment by MatthewFlamm** [comment]
I am +1 on both. I much prefer working with an auto formatter as it releases mental load to focus on other things, even if some cases make the code less readable.

I also have been using a git hook pre-commit from @adeak suggestion from my repeated formatting commits, but I don't like that it doesn't take into account the difference between the current workspace and the code to be committed.  I think python pre-commit may get around this issue, but I'm not sure.

pre-commit can also be opt-in (I think?), which lowers the barrier for other users.

**Reply by adeak** [reply]
> pre-commit can also be opt-in (I think?), 

Indeed, you have to pro-actively install `pre-commit` and then use it to install our hooks. This was the extra step I mentioned in my comment.

We can definitely make smarter hooks if we go down this path. My version is a least-effort solution.

**Reply by akaszynski** [reply]
> pre-commit can also be opt-in (I think?), which lowers the barrier for other users.

It can, and if you want to try it out without adding it as a hook, then run:
```
$ pre-commit run
black................................................(no files to check)Skipped
Reorder python imports...............................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
codespell................................................................Passed
pydocstyle...........................................(no files to check)Skipped
mypy.................................................(no files to check)Skipped
g[fix/final_edits f800b3e] skip venv check within workflow
 1 file changed, 1 insertion(+), 1 deletion(-)
```

The cool part is that it checks only staged files, so it's quite fast and it stops me from pushing some dumb commits.

As for implementing it within our CI/CD, that would indeed be a (minor) barrier to entry, but it's worth it. I'm a bit tired of fixing whitespace.

**Reply by adeak** [reply]
No, in CI it's fine. I specifically meant that "start contributing to pyvista" should be as lean as possible. If we add scary-looking steps like "install this", "run this to install a pre-commit hook" then it might scare some newbie contributors away. We can always help fix issues in CI.

**Reply by akaszynski** [reply]
As long as we make `pre-commit` optional but included in our documentation, it can actually reduce the barriers to entry as it's easier to validate that your PR is valid locally, especially on Windows.

**Reply by tkoyama010** [reply]
I'm still not familiar with pre-commit. If we `git commit` it without installing it, can we get a friendly message to install it?

**Reply by adeak** [reply]
> As long as we make pre-commit optional but included in our documentation, it can actually reduce the barriers to entry as it's easier to validate that your PR is valid locally, especially on Windows.

Indeed, that's what I had in mind. That's why I said that anything in pre-commit hooks should be optional.

> I'm still not familiar with pre-commit. If we git commit it without installing it, can we get a friendly message to install it?

My understanding is that git itself is only aware of installed hooks residing in `.git/hooks`, of which `.git/hooks/pre-commit` is one. So I doubt that `git commit` can give you a warning; that would also need a hook to be installed :) But we can include this step in our installation guide, so contributors only have to copy-paste some commands. So instead of "fork, clone, install, branch, work, commit, push" you'd have "fork, clone, install, install pre-commit, install hooks, branch, work, commit, push". Not really much additional work in practice, if one gets over the potentially daunting list of steps.

**Comment by banesullivan** [comment]
+1 for using Black with `skip-string-normalization=True` and something like `flake8-quotes` for enforcing the beloved single quote `'`

See what I've done in [`scooby`](https://github.com/banesullivan/scooby) to have black auto-formatting in https://github.com/banesullivan/scooby/pull/65

I'd be happy to set most of this up and make sure the CI handles the style checks

**Reply by banesullivan** [reply]
I started in #2180, because I've been eager for PyVista to use an auto-formatter... I'm partial to Black 😄
